### PR TITLE
Move `MeasureUnit` to separate module and Implement `UnitsTrieV1` Data Provider

### PR DESCRIPTION
**Pull request by younies**
## Task

Step 2 in refactoring `MeasureUnit` out of the `units` crate.

### Summary of Changes in this PR:

1. Moved `MeasureUnit` to the newly created `measure` crate.
2. Implemented a `UnitsTrieV1` data provider to support `MeasureUnit`.
3. Relocated auxiliary components, including `Si_Prefixes`.

### Next Steps:

1. Update `MeasureUnitParser` to utilize data directly from `UnitsTrieV1`.
2. Remove the units trie from `UnitsInfoV1`.
3. Develop a separate FFI for `MeasureUnitParser`.
4. Rename the `measure` crate to `icu_units_bikeshed` following the migration.

### Related Issues:

- https://github.com/unicode-org/icu4x/issues/5335

### CODE REVIEW

File: components/experimental/src/units/converter_factory.rs [hunk 1]
```diff
@@ -2,15 +2,16 @@
 // called LICENSE at the top level of the ICU4X source tree
 // (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
 
-use crate::units::provider;
-use crate::units::provider::MeasureUnitItem;
+use crate::measure::measureunit::MeasureUnit;
+use crate::measure::parser::MeasureUnitParser;
+use crate::measure::provider::single_unit::SingleUnit;
+use crate::units::provider::{self};
```
Comment (robertbastian) [inline components/experimental/src/units/converter_factory.rs:8]: does this pass `cargo fmt`?

```suggestion
use crate::units::provider;
```

File: components/experimental/src/measure/provider/single_unit.rs [hunk 2]
```diff
@@ -0,0 +1,28 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use super::si_prefix::SiPrefix;
+
+/// Represents a single unit in a measure unit.
+/// For example, the MeasureUnit `kilometer-per-square-second` contains two single units:
+///    1. `kilometer` with power 1 and prefix 3 with base 10.
+///    2. `second` with power -2 and prefix `NotExist`.
```
Comment (robertbastian) [inline components/experimental/src/measure/provider/single_unit.rs:10]: how can the prefix be `NotExist` if the `si_prefix` field is not optional?

File: components/experimental/src/units/provider.rs [hunk 3]
```diff
@@ -42,6 +43,7 @@ use super::ratio::IcuRatio;
 )]
 #[cfg_attr(feature = "serde", derive(serde::Deserialize))]
 pub struct UnitsInfoV1<'data> {
+    // TODO: remove this field once we are using this map from `measure/provider::UnitsTrie`.
```
Comment (robertbastian) [inline components/experimental/src/units/provider.rs:46]: does this mean with this PR there is data duplication?

File: provider/source/src/units/trie.rs [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        .load(DataRequest {
+            id: DataIdentifierCow::from_locale(langid!("und").into()).as_borrowed(),
+            ..Default::default()
+        })
+        .unwrap();
+
+    let units_info = und.payload.get().to_owned();
+    let trie = &units_info.trie;
+
+    assert_eq!(trie.get("meter").unwrap(), 84);
```
Comment (robertbastian) [inline provider/source/src/units/trie.rs:78]: Testing that meter is a particular index is extremely brittle and will break at each CLDR update: [CLDR 45 diff](https://github.com/unicode-org/icu4x/pull/4782/files#diff-3db58552c2b814e04560154124f3cceb7d37d8bc200ea517a09082fb1f354e1d). Ideally the indices shouldn't even show in JSON data and be entirely an encoding detail.

File: components/experimental/src/measure/provider/single_unit.rs (hunk 2 above)
Comment (younies) [reply components/experimental/src/measure/provider/single_unit.rs]: not exist means `Decimal` or `Binary` with  a power equal to `0`

We can say, and prefix power equal `0`

File: components/experimental/src/units/converter_factory.rs (hunk 1 above)
Comment (younies) [reply components/experimental/src/units/converter_factory.rs:8]: yes, but that is weird

Comment (younies) [reply components/experimental/src/units/converter_factory.rs:8]: done

File: components/experimental/src/units/provider.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/units/provider.rs:46]: yes, exactly

File: provider/source/src/units/trie.rs (hunk 4 above)
Comment (younies) [reply provider/source/src/units/trie.rs:78]: as you see in the TODO
```
    // TODO: add more tests that test the accessing of UnitsInfoV1Marker and get the correct info for each couple of units
```

I want in the next PR, to test that the id is pointing to the correct units info. After removing the duplication and use these data in the units converter. I will add these test cases.

Comment (robertbastian) [reply provider/source/src/units/trie.rs:78]: There is no guarantee to me as the reviewer that any TODO ever gets addressed. This is a blocking issue for me, because I will be the one who'll need to fix this during the next CLDR update if it's still there.

Either make this PR test that it's pointing to the correct units info, or remove this assertion.

File: components/experimental/src/units/provider.rs (hunk 3 above)
Comment (robertbastian) [reply components/experimental/src/units/provider.rs:46]: I don't like that you're adding so many TODOs in each PR. To me as a reviewer it's very hard to follow what is happening across multiple PRs, especially if I don't review all of them. I would prefer if you addressed the data duplication in this PR.

File: provider/source/src/units/trie.rs (hunk 4 above)
Comment (younies) [reply provider/source/src/units/trie.rs:78]: Okay, now I added a reasonable test cases that can proof why we need these ids.

File: components/experimental/src/units/provider.rs (hunk 3 above)
Comment (younies) [reply components/experimental/src/units/provider.rs:46]: The problem, if I finalized this PR with everything, it will be so huge that it is hard to be reviewed. I like to split the PRs to a reasonable amount of code.

**Review by robertbastian — APPROVED** [review]
